exonum-java-binding/qa-service/src/test/java/com/exonum/binding/qaservice/transactions/ErrorTxIntegrationTest.java

```diff
@@ -98,68 +122,41 @@ void constructorRejectsInvalidDescription() {
 
   @Test
   @RequiresNativeLibrary
-  void executeNoDescription() throws CloseFailuresException {
-    try (MemoryDb db = MemoryDb.newInstance();
-        Cleaner cleaner = new Cleaner()) {
-      Fork view = db.createFork(cleaner);
-
-      byte errorCode = 2;
-      Transaction tx = new ErrorTx(1L, errorCode, null);
-
-      TransactionContext context = newContext(view);
-      TransactionExecutionException expected = assertThrows(TransactionExecutionException.class,
-          () -> tx.execute(context));
-
-      assertThat(expected.getErrorCode(), equalTo(errorCode));
-      assertNull(expected.getMessage());
-    }
-  }
-
-  @Test
-  @RequiresNativeLibrary
-  void executeWithDescription() throws CloseFailuresException {
-    try (MemoryDb db = MemoryDb.newInstance();
-        Cleaner cleaner = new Cleaner()) {
-      Fork view = db.createFork(cleaner);
-
-      byte errorCode = 2;
-      String description = "Boom";
-      Transaction tx = new ErrorTx(1L, errorCode, description);
-
-      TransactionContext context = newContext(view);
-      TransactionExecutionException expected = assertThrows(TransactionExecutionException.class,
-          () -> tx.execute(context));
-
-      assertThat(expected.getErrorCode(), equalTo(errorCode));
-      assertThat(expected.getMessage(), equalTo(description));
+  void executeNoDescription() {
+    try (TestKit testKit = TestKit.forService(QaServiceModule.class)) {
+      byte errorCode = 1;
+      TransactionMessage errorTx = createErrorTransaction(0L, errorCode, null);
+      testKit.createBlockWithTransactions(errorTx);
+      List<String> logMessages = logAppender.getMessages();
+      // Logger contains two #getStateHashes messages and an exception message
+      int expectedNumMessages = 3;
+      assertThat(logMessages).hasSize(expectedNumMessages);
+
+      String exceptionMessage = "com.exonum.binding.transaction.TransactionExecutionException: null";
+      assertThat(logMessages.get(1))
+          .contains("INFO")
+          .contains(exceptionMessage);
     }
   }
 
   @Test
   @RequiresNativeLibrary
-  void executeClearsQaServiceData() throws CloseFailuresException {
```

Comment:
Here and below - as we throw inside `execute` method, exonum rollbacks any changes made (clearing schema in this case). So no, I don't see a way to replicate this test with TestKit, but I don't see how it is useful to us - we could check that it _does not_ clear schema though.

Reply:
It is not useful to test that exonum rollbacks in a usual service, but is in the qa service. It modifies data on purpose so that we can test exonum itself (= Java integration with exonum). Adding anything specific to testkit for such scenario does not make sense. But the original tests must remain to verify these transactions (using MemoryDb, if nothing else works).